Add support for IPv6 in Data Channel TCP

VirrageS opened the pull request:
closes #9

VirrageS committed:
Add support for IPv6 in Data Channel TCP

apaszke commented on torch/lib/THD/base/channels/DataChannelTCP.cpp:
```diff
@@ -195,49 +195,60 @@ DataChannelTCP::~DataChannelTCP()
-  addr.sin_family = AF_INET;
-  addr.sin_port = htons(port);
-  addr.sin_addr.s_addr = INADDR_ANY;
+  // TODO: should we prioritize any `ai_family`?
```
How would you like to prioritize it? First try with IPv6 and fallback to IPv4?

VirrageS replied:
Yes. `res` is list with resolved (found) IPv6 and IPv4 addresses, so we can iterate through list checking if address has `AF_INET6` family. Then try to make socket. If it fails we can try to find another one IPv6 or try IPv4 addresses if there is no IPv6 left and try to make socket once again (until succeeded or out of addresses).

apaszke commented on torch/lib/THD/base/channels/DataChannelTCP.cpp:
```diff
@@ -195,49 +195,60 @@ DataChannelTCP::~DataChannelTCP()
+  struct addrinfo hints, *res = NULL;
+
+  memset(&hints, 0x00, sizeof(hints));
+  hints.ai_flags = AI_NUMERICSERV;
```
What's the flag?

VirrageS replied:
This flag says that second parameter of `getaddrinfo` (which is called `service`) should be interpreted as decimal port number (not binary).

"If AI_NUMERICSERV is specified in hints.ai_flags and service is not NULL, then service must point to a string containing a numeric port number"

apaszke commented on torch/lib/THD/base/channels/DataChannelTCP.cpp:
```diff
@@ -195,49 +195,60 @@ DataChannelTCP::~DataChannelTCP()
+      hints.ai_family = AF_INET6;
+      hints.ai_flags |= AI_NUMERICHOST;
+    }
+  }
```
Can you briefly explain to me what happens in here?

VirrageS replied:
So `hints` struct is struct which can be used to narrow search and return range of `getaddrinfo` which will try to find addresses linked to provided host and port.

So to narrow it even more we can try to check if host is actually IP (either IPv4 text address or IPv6 text address). So firstly we try to check if host is valid IPv4 address and if it fails IPv6 address then set flags respectively.

This code is **not** necessary, it just speed up (not sure how much and if at all) resolving of `getaddrinfo` function and narrow set of returned addresses which result in less memory allocation/consumption.

VirrageS committed:
Add multiple addresses checking
- Removed IPv4 or IPv6 address checking before `getaddrinfo`. It
was unnecessary and only caused confusion
- Abandoned idea with prioritizing IPv6. It occured that
`getaddrinfo` sort addresses according to better algorithm
- Added comments

apaszke commented on torch/lib/THD/base/channels/DataChannelTCP.cpp:
```diff
@@ -196,70 +196,99 @@ DataChannelTCP::~DataChannelTCP()
+  hints.ai_family = AF_UNSPEC; // either IPv4 or IPv6
+  hints.ai_socktype = SOCK_STREAM; // TCP
+
+  // `getaddrinfo` will sort addresses according to RFC 3484 and can be tweeked
```
what's the default sorting strategy?

VirrageS replied:
Of `getaddrinfo`? It's quite long and consists of numbers of rules: https://github.com/lattera/glibc/blob/master/sysdeps/posix/getaddrinfo.c#L1591

I have decided to remove IPv6 preference over IPv4 and just stick to RFC which I guess is thought out. So instead greedy IPv6 addresses picking we get more carefully picked addresses.

And as it is pointed out in comment, when someone wants to change the order, it just change order in some specified file `/etc/gai.conf` instead of changing code (which is not that easy for a normal user).

VirrageS replied:
The problem arise when we want to change this file (`/etc/gai.conf`) on another system different than Linux because it usually does not exist or exists under a different name. Not sure how it is implemented in OSX or other UNIX system. So I am not 100% about abandoning IPv6 preference.

So if we want still have IPv6 preference over IPv4 I can change that in a minute. But as I pointed out then we give up possibly better sorting to greedy picking.

apaszke replied:
I think we can just leave it as is, maybe it will work out fine. I just know that I wouldn't be happy having to tell users that they should change a system-wide config file to make it work 😉

apaszke commented on torch/lib/THD/base/channels/DataChannelTCP.cpp:
```diff
@@ -196,70 +196,99 @@ DataChannelTCP::~DataChannelTCP()
+  // get listen port
+  struct sockaddr_in addr;
+  socklen_t addr_len = sizeof(addr);
+  SYSCHECK(::getsockname(_socket, reinterpret_cast<struct sockaddr*>(&addr), &addr_len))
```
If this fails, you'll leak `res`

apaszke replied:
It would be simpler if you wrote a `Deleter` class for `addrinfo*` that calls `freeaddrinfo`. You could use a smart pointer and not worry about leaks then.

VirrageS committed:
Fix possible memory leak
Added smart pointer (shared_ptr) to handle freeing `addrinfo` structure.
Now any exception will not cause memory leak.

jytug commented on torch/lib/THD/base/channels/DataChannelTCP.cpp:
```diff
@@ -196,67 +196,99 @@ DataChannelTCP::~DataChannelTCP()
+  // by editing `/etc/gai.conf`. so there is no need to manual sorting
+  // or protcol preference.
+  int err = getaddrinfo(NULL, std::to_string(port).data(), &hints, &res);
+  if (err != 0 || !res) {
```
Should this be tested by `SYSCHECK`?

VirrageS replied:
It could but changing that would involve some unnecessary code since we want to print `gai_strerror`. To do that we would need to do:
```
int err;
try {
  SYSCHECK(...)
} catch (system_error) {
  invalid_argument(... gai_strerror)
}

if (!res) {
  invalid_argument(...)
}
```

jytug commented on torch/lib/THD/base/channels/DataChannelTCP.cpp:
```diff
@@ -196,67 +196,99 @@ DataChannelTCP::~DataChannelTCP()
+  // `getaddrinfo` will sort addresses according to RFC 3484 and can be tweeked
+  // by editing `/etc/gai.conf`. so there is no need to manual sorting
+  // or protcol preference.
+  int err = ::getaddrinfo(address.data(), std::to_string(port).data(), &hints, &res);
```
I think it's better to use `c_str()` instead of `data()` here, since it gives you a null-terminated string

VirrageS replied:
http://stackoverflow.com/a/12399518:

> In C++11/C++0x, data() and c_str() is no longer different. And thus data() is required to have a null termination at the end as well.
